Create 'disableReload' config option to handle infinite reload case

messman opened the pull request:
From #1353:

> For a full solution, I would suggest the addition of an option in `cosmos.config.json` for controlling the `reload` behavior of HMR. I'd like to know what others think. Cheers.

> In the meantime I like your idea to add a Cosmos config option to disable the reload flag. Would you be willing to make this PR?

This PR adds a new Cosmos config option `disableReload` that, when `true`, sets the internal webpack-hot-middleware `reload` option to `false`. This, in turn, causes the hot reload code to _not_ do a `window.location.reload` when an error is encountered (no modules to update; error while updating modules; etc) by webpack during hot reload.

This config option is not required and defaults to `false`, and as such is not a breaking change.

The option is named `disableReload` for brevity and to match the naming convention for an existing option, `disablePlugins`. "disableReload" is a potentially confusing name, as one may see it as controlling the overall hot reload experience; however, this use of the term "reload" to only apply to a _window location reload_ is consistent with the naming of the `reload` option in webpack-hot-middleware. `reload` controls unexpected update behaviors, not expected (happy-path) "update" ones.

Changes include:
- Updating cosmos config schema (will need to be updated on schemastore)
- Adding TypeScript property
- Updating function responsible for getting the hot middleware configuration string

Tested and fixes issue in #1353 as first described in its [test repo](https://github.com/messman/cosmos-reload-example).

skidding commented:
Hi @messman,

This is great. I love how detailed your PR description is!

How about putting the `disableReload` option inside the `webpack` object config?
It makes sense because it's webpack-related and we get to keep this option away from the Cosmos core. I won't get into the (undocumented...) Cosmos plugin system here but in your case the logic where you provide the default value for the `webpack.disableReload` option would be moved to here: [packages/react-cosmos/src/plugins/webpack/cosmosConfig/webpack.ts](https://github.com/react-cosmos/react-cosmos/blob/333dfd904a2e9d99b20fb086f984b1d21a43ebd9/packages/react-cosmos/src/plugins/webpack/cosmosConfig/webpack.ts).

messman commented:
> How about putting the `disableReload` option inside the `webpack` config object?

Ah yes, that makes sense. I see how keeping it close to `hotReload` is beneficial. I have updated the code.

I also modified the property to be `reloadOnFail` with a default of `true` rather than `disableReload` with a default of `false`. This is in spite of the argument I made earlier about consistency with webpack-hot-middleware. In my view, `hotReload` and `disableReload` as siblings inside the `webpack` config object would be confusing, and `reloadOnFail` is a better descriptor of the behavior. If you disagree, I can easily change it back to the prior discussed `disableReload` or use a different name. 👍

skidding commented:
`reloadOnFail` is even better.

Published a patch with this change in `react-cosmos@5.6.5`. Please try it out at let me know if it works as expected.

Thanks!

messman commented:
@skidding:

`react-cosmos@5.6.5` does have the fix and it works as expected. 🎉  Thanks!

Since the `cosmos.config.json` schema was updated, http://json.schemastore.org/cosmos-config needs to be updated as well. Is that baked in to the deploy process or is it manual?
Once that's handled everything is set.

skidding commented:
Updating the schema store is manual. Here are some PRs from the past: https://github.com/SchemaStore/schemastore/pulls?q=is%3Apr+author%3Askidding+is%3Aclosed – Do you want to do it?

messman commented:
> Do you want to do it?

Sure, not a problem. Thanks for the PR list – it was helpful to see the pattern.

Created SchemaStore/schemastore#1966, which also contains changes related to #1299 brought in from the copy.
